db/version_builder.cc

```diff
@@ -382,43 +365,76 @@ class VersionBuilder::Rep {
     ExpectedLinkedSsts expected_linked_ssts;
 
     if (num_levels_ > 0) {
+      const InternalKeyComparator* const icmp = vstorage->InternalComparator();
+      EpochNumberRequirement epoch_number_requirement =
+          vstorage->GetEpochNumberRequirement();
+      assert(icmp);
       // Check L0
       {
-        auto l0_checker = [this](const FileMetaData* lhs,
-                                 const FileMetaData* rhs) {
+        auto l0_checker = [this, epoch_number_requirement, icmp](
+                              const FileMetaData* lhs,
+                              const FileMetaData* rhs) {
           assert(lhs);
           assert(rhs);
 
-          if (!level_zero_cmp_(lhs, rhs)) {
-            std::ostringstream oss;
-            oss << "L0 files are not sorted properly: files #"
-                << lhs->fd.GetNumber() << ", #" << rhs->fd.GetNumber();
+          if (epoch_number_requirement ==
+              EpochNumberRequirement::kMightMissing) {
+            if (!level_zero_cmp_by_seqno_(lhs, rhs)) {
+              std::ostringstream oss;
+              oss << "L0 files are not sorted properly: files #"
+                  << lhs->fd.GetNumber() << " with seqnos (largest, smallest) "
+                  << lhs->fd.largest_seqno << " , " << lhs->fd.smallest_seqno
+                  << ", #" << rhs->fd.GetNumber()
+                  << " with seqnos (largest, smallest) "
+                  << rhs->fd.largest_seqno << " , " << rhs->fd.smallest_seqno;
+              return Status::Corruption("VersionBuilder", oss.str());
+            }
+          } else if (epoch_number_requirement ==
+                     EpochNumberRequirement::kMustPresent) {
+            if (lhs->epoch_number == kUnknownEpochNumber) {
+              std::ostringstream oss;
+              oss << "L0 file is not assigned with valid epoch number: files #"
+                  << lhs->fd.GetNumber() << " with epoch number "
+                  << lhs->epoch_number;
+              return Status::Corruption("VersionBuilder", oss.str());
+            }
 
-            return Status::Corruption("VersionBuilder", oss.str());
-          }
+            if (rhs->epoch_number == kUnknownEpochNumber) {
+              std::ostringstream oss;
+              oss << "L0 file is not assigned with valid epoch number: files #"
+                  << rhs->fd.GetNumber() << " with epoch number "
+                  << rhs->epoch_number;
+              return Status::Corruption("VersionBuilder", oss.str());
+            }
```

Comment:
As mentioned in another comment, there may be a better place (write side) for this validation.

Reply:
Fixed by removing this kUnknownEpochNumber and revised the related UT.